require(rjson)
require(dplyr)

####################################################################
### Add rsids to a file annotated with only chromosome and position
####################################################################

# Required config parameters (if this step is not skipped):
#
# "output_dir": "dir4"
# Path to the directory where all output files should be placed, including
# those generated by earlier steps in this post-processing
#
# "pre_coloc_test_list": "test_list_file"
# A file containing a list of all the tests to be run through the colocalization
# pipeline. This QC script will check to make sure they were all properly
# executed.

# Optional config parameters:
#
# "skip_steps": ["coloc_qc", ...]
# If "coloc_qc" is not included in the "skip_steps" list
# parameter, then it will run by default. If included,
# then this step will be skipped.
#
# "coloc_qc" :
# {
# 	"input_results_file": "file1",		# Optional
# 	"output_qc_file": "file2",		# Optional
#
#	"dbsnp_file": "map-file",		# Optional
#	"rsid_col_index": 1,			# Required if "dbsnp_file" is specified
#	"chr_col_index": 2,			# Required if "dbsnp_file" is specified
#	"pos_col_index": 3,			# Required if "dbsnp_file" is specified
#	"use_tabix": "FALSE"			# Recommended if not using a tabix'ed dbSNP file (the default file is tabix'ed)
# }
#
# If included, "add_hgnc_names" is a JSON object that contains additional parameters
# that will be used for running this filtering step.
#
#	"input_results_file" is, optionally, the location of the colocalization results
#	for which QC should be run. If not specified, the output from the previous step
#	in the pipeline will be used by default, namely TODO:
#
#	"output_qc_file" is, optionally, the file path to which a QC summary should be
#	written. If not specified, the QC summary will be written to the default location,
#	namely TODO:
#
#	# TODO: Add this to the pipeline flowchart after the categorization step, as an
#	essential step in the process

main = function()
{
	# Load config file

	# Load final results file AND pre-coloc test list

	# Give each test a test-ID, something like this
	pre_run_coloc$test_id = paste(pre_run_coloc$chr, pre_run_coloc$snp_pos, pre_run_coloc$eqtl_file, pre_run_coloc$gwas_file, pre_run_coloc$feature, sep="_")
	post_run_coloc$test_id = paste(post_run_coloc$ref_snp, post_run_coloc$eqtl_file, post_run_coloc$base_gwas_file, post_run_coloc$feature, sep="_")

	# Load the list of filters that were performed; that way we know not to worry if
	# tests didn't show up simply because they were removed in post-hoc filtering

	# Remove all tests from the pre-coloc list if they shouldn't be there
	pre_run_coloc = remove_post_hoc_exemptions(pre_run_coloc, config)

	# Print numbers of loci tested
	quantify_successful_tests()

	# Sanity check GWAS and eQTL p-values
	# TODO: Print any where none were run...
	verify_gwas_ranges_and_counts()
	verify_eqtl_ranges_and_counts()

	# Make sure all tested sites had a reasonable total number of SNPs for colocalization testing
	verify_snp_density()
}


remove_post_hoc_exemptions = function(pre_run_coloc, config)
{
	# TODO: Make this general
	pre_run_coloc = pre_run_coloc[!grepl("2hr", pre_run_coloc$gwas_file),]
	pre_run_coloc = pre_run_coloc[(grepl("ISI_Model_", pre_run_coloc$gwas_file) | grepl("MI_adjBMI_", pre_run_coloc$gwas_file) | (pre_run_coloc$gwas_pvalue < 5e-8)),]
}

quantify_successful_tests = function()
{
	print("Number of candidate loci successfully tested:")
	sum(pre_run_coloc$test_id %in% post_run_coloc$test_id)
	print("")
	print("Number of candidate loci not appearing in final results:")
	sum(!(pre_run_coloc$test_id %in% post_run_coloc$test_id))
	print("")

	log_rejected_tests()
}

log_rejected_tests = function(x)
{
	rejected_tests = pre_run_coloc[!(pre_run_coloc$test_id %in% post_run_coloc$test_id),]
}


verify_gwas_ranges_and_counts = function()
{
	# Make sure the range of GWAS p-values for each type of test is reasonable

	# Visual inspection:
	# Each GWAS should have a reasonable number of tested SNPs.
	# The range of SNP pvalues for each GWAS study should match the specified cutoffs
	print("Number of tested SNPs and p-value range for each GWAS:")
	gwas_stats = post_run_coloc %>% group_by(base_gwas_file) %>% summarize(total_snps=length(unique(ref_snp)), gwas_range=paste(round(range(X.log_gwas_pval),2), collapse="..."), eqtl_range = paste(round(range(X.log_eqtl_pval),2), collapse="..."))
	print(gwas_stats)
	print("")
}

verify_eqtl_ranges_and_counts = function()
{
	# Make sure the range of eQTL p-values for each type of test is reasonable

	# Visual inspection:
	# Each eQTL should have a reasonable number of tested SNPs.
	# The range of SNP pvalues for each eQTL study should match the specified cutoffs

	print("Number of tested SNPs and p-value range for each QTL study:")
	eqtl_stats = post_run_coloc %>% group_by(eqtl_file) %>% summarize(total_snps=length(unique(ref_snp)), gwas_range=paste(round(range(X.log_gwas_pval),2), collapse="..."), eqtl_range = paste(round(range(X.log_eqtl_pval),2), collapse="..."))
	print(eqtl_stats)
	print("")
}

verify_snp_density = function()
{
	# We shouldn't see many sites with < 50 or so SNPs. We shouldn't really see any
	# with < 10 SNPs; if we do, we need to toss these sites at an earlier step in the
	# process.
	# TODO: Add a min_snps parameter to the pre-run config file
	print("Range of number of SNPs per locus:")
	print(range(post_run_coloc$n_snps))
	print("")

	# The violators are all FastGlu and FastInsu SNPs, which isn't surprising.
	# This is a good sanity check though.
	print("Loci with low number of SNPs:")
	print(table(post_run_coloc[post_run_coloc$n_snps < 50,]$base_gwas_file))
	print("")
}

main()
