require(rjson)
require(dplyr)

####################################################################
### Add rsids to a file annotated with only chromosome and position
####################################################################

# Required config parameters (if this step is not skipped):
#
# "output_dir": "dir4"
# Path to the directory where all output files should be placed, including
# those generated by earlier steps in this post-processing
#
# "genome_build": "hg19"
# The desired build of the genome for matching rsids. (hg19 or hg38).
# If a custom rsid mapping file is supplied (see optional parameters section below),
# then this parameter can be omitted.

# Optional config parameters:
#
# "skip_steps": ["add_rsids", ...]
# If "add_rsids" is not included in the "skip_steps" list
# parameter, then it will run by default. If included,
# then this step will be skipped.
#
# "add_rsids" :
# {
# 	"input_results_file": "file1",		# Optional
# 	"output_results_file": "file2",		# Optional
#
#	"dbsnp_file": "map-file",		# Optional
#	"rsid_col_index": 1,			# Required if "dbsnp_file" is specified
#	"chr_col_index": 2,			# Required if "dbsnp_file" is specified
#	"pos_col_index": 3,			# Required if "dbsnp_file" is specified
#	"use_tabix": "FALSE"			# Recommended if not using a tabix'ed dbSNP file (the default file is tabix'ed)
# }
#
# If included, "add_hgnc_names" is a JSON object that contains additional parameters
# that will be used for running this filtering step.
#
#	"input_results_file" is, optionally, the location of the colocalization results file to which the 
# 	rsids should be added. This file should be in TSV format and must
#	contain a column named "ref_snp" that contains the {chr}_{pos} of all SNPs for which
#	rsids will be retrieved. If "input_results_file" is not included as a parameter, the default
#	will be to load the output from the "assign_locus_numbers" step, namely the file at
#	{output_dir}/filtered_coloc_table_with_loci.txt.
#	
#	"output_results_file" is, optionally, the file to which the output of this step
#	should be written. If not specified, it will be placed at
#	{output_dir}/filtered_coloc_with_loci_hgnc.txt.
#
#	"dbsnp_file", is, optionally, a TSV with a mapping from chromosomal coordinates to rsids.
#	If specified, then there are TWO POSSIBLE types for this file, each of which requires 
#	that certain additional parameters be specified:
#	
#		Format 1: Tabix'ed, bgzipped file (recommended)
#		If a tabix'ed, bgzipped file is provided, the script will run much faster and will
#		not consume much memory; thus, we recommend this option whenever possible.
#			
#			If this format is chosen, you must additionally specify an "rsid_col_index" (the
#			1-based index of the column containing rsids). Also, there must exist a tabix index file in
#			the same location as "dbsnp_file" with the name `{dbsnp_file}.tbi`.
#
#			If you're unfamiliar with bgzip and tabix, we recommend you check out the tool at
#			http://www.htslib.org/doc/tabix.html ; this package is quite easy to use and might
#			save you hours, days, even weeks of computation time in the future :)
#
#			If the above explanation sounds entirely mystifying, we recommend you just use the
#			default settings instead of supplying a custom "dbsnp_file"; this will produce reasonable
#			rsids and won't have any impact on subsequent tools in this toolkit.
#
#		Format 2: Regular TSV file.
#		This format is typically slower because it requires loading the entire SNP-to-rsid mapping into memory.
#		Thus, if using this option, please ensure that your system has enough memory to load the entire file into
#		memory, or it may slow down your system considerably.
#			
#			If this option is chosen, you must supply "rsid_col_index", "chr_col_index", "pos_col_index" 
#			specifying the 1-indexed column indices of the rsid, chromosome,
#			and position on the chromosome respectively. If "dbsnp_file" is NOT specified, then the user must
#			instead specify a "genome_build" parameter (see required parameters section above). You should also
#			supply the parameter "tabix": "FALSE", since by default this script will attempt to use tabix to
#			fetch the required rsids.
#
#	TODO: Add a step of cross-referencing the ref / alt alleles with those in dbSNP to be totally sure they match
#

default_hg38_dbsnp = "data/dbsnp/hg38/common_all_20170710.vcf.gz"
default_hg19_dbsnp = "data/dbsnp/hg19/common_all_20170710.vcf.gz"
default_chr_col_index = 1
default_pos_col_index = 2
default_rsid_col_index = 3

add_rsids = function(config_file, input_file, output_file)
{
	config = fromJSON(file=config_file)$add_rsids
	config$input_file = input_file
	config$output_file = output_file

	# Load results table
	results = load_results_file_for_rsids(config)
	# Add rsids, probably by tabix lookup...
	results = add_rsid_column(results, config)

	# Write output table of results with rsids
	print(head(results))
	write.table(results, config$output_file, col.names=TRUE, row.names=FALSE, quote=FALSE, sep="\t")
}

add_rsid_column = function(results, config)
{
	# NOTE: This might be expanded later to include custom dbSNP file; for now
	# we're just assuming the default files
	if (FALSE)
	{
		if (("use_tabix" %in% names(config$add_rsids)) && (config$use_tabix == "FALSE"))
		{
			# If explicitly told the custom file is not tabix'ed, load it to memory
			return(get_rsids_without_tabix(results, config$dbsnp_file, config$rsid_col_index,
						       config$chr_col_index, config$pos_col_index))	
		} else if (file.exists(paste0(config$dbsnp_file, ".tbi")))
		{
			# Tabix file exists, so use it
			return(get_rsids_with_tabix(results, config$dbsnp_file, config$rsid_col_index))
		} else
		{
			print("Warning: No tabix file exists for 'add_rsids'; custom SNP-to-rsid 
			      map will now be fully loaded into memory.")	
			# Tabix file doesn't even exist, so we're loading the file to memory
			return(get_rsids_without_tabix(results, config$dbsnp_file, config$rsid_col_index,
						       config$chr_col_index, config$pos_col_index))	
		}
	}

	if (config$genome_file == "hg19")
	{
		return(get_rsids_with_tabix(results, default_hg19_dbsnp, default_rsid_col_index,
			default_chr_col_index, default_pos_col_index))
	} else if (config$genome_file == "hg38")
	{
		return(get_rsids_with_tabix(results, default_hg38_dbsnp, default_rsid_col_index))
	} else
	{
		stop('Must specify genome build in the config file:
		     set config -> add_rsids -> genome_file to either "hg19" or "hg38"')
	}
}

get_rsids_with_tabix = function(results, dbsnp_file, rsid_index)
{
	# Load rsIDs -- here, we'll use a simple lookup in the 1Kgenomes file
	rsids = results[!duplicated(results$ref_snp),][c("ref_snp")]
	rsids$chr = sapply(as.character(rsids$ref_snp), function(x) {strsplit(x, "_")[[1]][1]})
	rsids$pos = sapply(as.character(rsids$ref_snp), function(x) {strsplit(x, "_")[[1]][2]})
		
	rsids$rsid = sapply(1:dim(rsids)[1], function(i)
	{
		print(i)
		chr = rsids$chr[i]
		pos = rsids$pos[i]
		# Try it with and without the "chr" notation at the front
		query = c(system(paste0("tabix ", dbsnp_file, " chr", chr, ":", pos, "-", pos), intern=TRUE),
			system(paste0("tabix ", dbsnp_file, " ", chr, ":", pos, "-", pos), intern=TRUE))
		if (length(query) > 1)
		{
			vars = sapply(1:length(query), function(i) {strsplit(query[[i]], "\\t")[[1]][rsid_index]})
			vars2 = vars[grepl("rs", vars)]
			if (length(vars2) > 0)
			{
				return(paste(vars2, collapse=","))
			}
			return(paste(vars, collapse=","))
		}
		if (length(query) == 0)
		{
			# No rsid variant found in the map
			return(NA)
		}
		rsid = strsplit(query, "\\t")[[1]][rsid_index]
		return(rsid)
	})

	return(merge(results, rsids))
}

get_rsids_without_tabix = function(results, dbsnp_file, rsid_index, chr_index, pos_index)
{
	rsids = read.table(dbsnp_file, header=FALSE, fill=TRUE, sep="\t")
	rsids = rsids[,c(rsid_index, chr_index, pos_index)]
	colnames(rsids) = c("rsid", "chr", "pos")
	rsids$ref_snp = paste(rsids$chr, rsids$pos, sep="_")
	
	rsids %>% group_by(ref_snp) %>% summarize(rsid = paste(rsid, collapse=","))
	return(merge(results, rsids))
}

validate_config = function(config_file)
{
        # Load config file, specified as a command line parameter
        config = fromJSON(file=config_file)

	# Make sure the columns are marked if a custom rsid file is being used.
        if (("add_rsids" %in% names(config)))
        {
		if (("dbsnp_file" %in% names(config$add_rsids)))
		{
			if (!("rsid_col_index" %in% names(config$add_rsids)))
			{
				stop("Config ERROR: When using a custom rsid mapping, you
				     must specify 'rsid_col_index' in the config file.")
			}
			if (!(("rsid_col_index" %in% names(config$add_rsids)) && ("chr_col_index" %in% names(config$add_rsids)
										  && ("pos_col_index" %in% names(config$add_rsids)))))
			{
				if (("use_tabix" %in% names(config$add_rsids)) && (config$add_rsids$use_tabix == "FALSE"))
				stop("Config ERROR: When using a custom rsid mapping that is not tabix'ed, you
				     must specify 'rsid_col_index', 'chr_col_index', and 'pos_col_index' in the config file")
			}

			if ("genome_build" %in% names(config))
			{
				print("Warning: you have specified both the 'genome_build' and the 'dbsnp_file' 
				      parameters. This is allowed, but the 'genome_build' will be ignored for this
				      step and the custom rsid file used instead, regardless of its genome version.'")
			}
		}
	}
	
	# Make sure the genome build is specified if no custom rsid file has been supplied.
	if ((!("add_rsids" %in% names(config))) || (!("dbsnp_file" %in% names(config$add_rsids))))
	{
		if (!("genome_build" %in% names(config)))
		{
			stop("Config ERROR: must either specify an rsid mapping file, or
			     specify the 'genome_build' parameter")	
		}
		else if (!(config$genome_build %in% c("hg19", "hg38")))
		{
			stop("Config ERROR: 'genome_build' parameter for 'add_rsids' must equal 
			     either 'hg19' or 'hg38'")	
		}

	}

	if (!("output_dir" %in% names(config)))
	{
	        stop("Config ERROR: You must specify 'output_dir' in config file")
	}

        return(config)
}

load_results_file_for_rsids = function(config)
{
	data = read.table(config$input_file, header=TRUE, sep="\t")

	if (!("ref_snp" %in% colnames(data)))
	{
		stop("input error: the input coloc results table for 'add_rsids' must have 
		     a 'ref_snp' column indicating chromosomal coordinates")
	}

	return(data)
}

args = commandArgs(trailingOnly=TRUE)

config_file = args[1]
input_file = args[2]
output_file = args[3]
add_rsids(config_file, input_file, output_file)
