require(rjson)

############################################################
### Filter colocalization results
############################################################

# Required config parameters (if this step is not skipped):
#
# "output_dir": "dir4"
# Path to the directory where all output files should be placed, including
# those generated by earlier steps in this post-processing

# Optional config parameters:
#
# "skip_steps": ["post-hoc-filter", ...]
# If "post-hoc-filter" is not included in the "skip_steps" list
# parameter, then it will run by default. If included,
# then this step will be skipped.
#
# "post-hoc-filter" :
# {
# 	"full_results_file": "results_file",	# Optional
# 	"out_results_file": "out_results_file",	# Optional
#
#	"kept_gwas": ["gwas1", "gwas2", ...],	# Optional ; CANNOT include both "kept_gwas" and "removed_gwas"
#	"kept_eqtl": ["eqtl1", "eqtl2", ...],	# Optional ; CANNOT include both "kept_eqtl" and "removed_eqtl"
#
#	"removed_gwas": ["gwas3"],		# Optional
#	"removed_eqtl": ["eqtl4"],		# Optional
#
#	"gwas_pval_threshold":			# Optional
#	{
#		"standard": 5e-8,		# Not optional if "gwas_pval_threshold" is included
#		"exceptions":			# Always optional
#		{
#			"special-gwas1": 1e-5,
#			"special-gwas2": 1e-5
#		}
#	},
#	"eqtl_pval_threshold":			# Optional
#	{
#		"standard": 1e-5,		# Not optional if "eqtl_pval_threshold" is included
#		"exceptions": {}		# Always optional
#	}
#
# }
# If included, "post-hoc-filter" is a JSON object that contains additional parameters
# that will be used for running this filtering step.
#
#	"full_results_file", "full_errors_file", and "full_skips_file" are file locations of
#	the results, errors, and skipped variants files, in the format produced by the "cat-results"
#	tool, relative to the top directory of this project.
#
#	If the cat-results tool was run prior to this tool with default parameters, then
#	these parameters don't need to be included, as the file in question will have been generated
#	in "output_dir" in a fixed location. Namely, the output file will be at {output_dir}/raw_coloc_table.txt.
#
#	"out_results_file" is the location where the results file should be written after filtering, 
#	relative to the top directory of this project. If not specified, it will be written by default 
#	to {output_dir}/filtered_coloc_table.txt.
#
#	"kept_gwas" is, optionally, a list of GWAS traits that should be included while all others are
#	filtered out. "removed_gwas", by contrast, is a list of GWAS traits that should be removed and
#	all others kept. To avoid conflicts, no more than one of "kept_gwas" and "removed_gwas" can be
#	specified in the options. "kept_eqtl" and "removed_eqtl" apply in a similar way. The format of the
#	GWAS and eQTL trait names in these fields should be identical to the strings used in the "eqtl_file"
#	and "trait" columns of the input files being filtered.
#
#	"gwas_pval_threshold" is, optionally, a p-value threshold at or above which results should be removed
#	from the results table. It must always include at least one field, "standard", which specifies a floating-
#	point number to use as the default threshold for filtering. Then, an "exceptions" field can optionally
# 	be included, to designate specific GWAS that are exceptions to the normal filtering threshold. In the example
#	above, "special-gwas1" and "special-gwas2" have been given exceptions to give them a less stringent GWAS cutoff
#	threshold (1e-5 instead of the standard 5e-8 that will be used for the rest of the loci). "eqtl_pval_threshold"
#	is exactly analogous to "gwas_pval_threshold".
#
#		A typical use case for this: perhaps you ran colocalization for 10 traits at all loci meeting a 
#		loose threshold of GWAS p-value < 1e-5. In your final presentation of the results, however, you'd
#		like to focus only on results meeting the typical genome-wide signficance threshold of GWAS
# 		p-value < 5e-8. But perhaps there are also 3 traits of particular interest, for which you're
#		willing to dig through a few extra, less-promising loci to ensure that you don't miss an interesting
#		ones. You set an exception for these 3 traits to include all loci passing the 1e-5 cutoff.
#
#		We note that there is not yet a rigorous statistical approach for controlling the false positive
# 		rate in these colocalization trials; therefore, care should be taken to follow up all identified
# 		colocalizations and to consider them within the context of other available multi-omic data. 
# 		In general, however, lowering the p-value cutoff detects more possible causal colocalizations, at
#		the cost of also turning up more likely false positives.
#
#	MAYBE also include other raw files here for filtering too while we're at it...



main = function()
{
	# Load config file
	config_file = commandArgs(trailingOnly=TRUE)[1]
	config = validate_config(config_file)

	# Load results, errors, skips files
	results = load_results_file(config)

	pre_results_dim = dim(results)[1]

	if ("post-hoc-filter" %in% names(config))
	{

		# If specified, filter results down to a limited set of GWAS and/or eQTL studies 

		if ("kept_gwas" %in% names(config$post-hoc-filter))
		{
			results = filter_by_gwas(results, config$kept_gwas, keep=TRUE)
		}
		else if ("removed_gwas" %in%  names(config$post-hoc-filter))
		{
			results = filter_by_gwas(results, config$removed_gwas, keep=FALSE)
		}

		if ("kept_eqtl" %in% names(config$post-hoc-filter))
		{
			results = filter_by_eqtls(results, config$kept_eqtls, keep=TRUE)
		}
		else if ("removed_eqtl" %in%  names(config$post-hoc-filter))
		{
			results = filter_by_gwas(results, config$removed_eqtls, keep=FALSE)
		}

		# Filter by p-vals
		results = apply_pval_filter(results, config)
	}

	# Display warning if not a single result was removed.
	if (dim(results)[1] == pre_results_dim)
	{
		print("Warning: No tests were removed during filtering.")
	}

	# Write filtered output files
	if (("post-hoc-filter" %in% names(config)) && ("out_results_file") %in% names(config$post-hoc-filter))
	{
		write.table(results, config$post-hoc-filter$out_results_file, quote=FALSE, sep="\t", col.names=TRUE, row.names=FALSE)
	} else
	{
		write.table(results, file=paste(config$out_dir, "filtered_coloc_table.txt", sep="/"), quote=FALSE, sep="\t", col.names=TRUE, row.names=FALSE)
	}
}

validate_config = function(config_file)
{
	# load config file, specified as a command line parameter
	config = fromjson(file=config_file)

	# validate config file to be sure required parameters are present
	if (!("output_dir" %in% names(config)))
	{
		stop("config error: you must specify 'output_dir' in config file")
	}

	if (("kept_gwas" %in% names(config)) && ("removed_gwas" %in%  names(config)))
	{
		stop("config error: you can't specify both 'kept_gwas' and 'removed_gwas' in config file")
	}
	
	if (("kept_eqtl" %in% names(config)) && ("removed_eqtl" %in%  names(config)))
	{
		stop("config error: you can't specify both 'kept_eqtl' and 'removed_eqtl' in config file")
	}
	if ((("post-hoc-filter" %in% names(config)) && ("gwas_pval_threshold" %in% names(config$post-hoc-filter)) && 
	     !("standard" %in% config$post-hoc-filter$gwas_pval_threshold)))
	{
		stop("config error: must include a 'standard' gwas threshold if 'gwas_pval_threshold' is specified.")
	}
	if ((("post-hoc-filter" %in% names(config)) && ("eqtl_pval_threshold" %in% names(config$post-hoc-filter)) && 
	     !("standard" %in% config$post-hoc-filter$eqtl_pval_threshold)))
	{
		stop("config error: must include a 'standard' eqtl threshold if 'eqtl_pval_threshold' is specified.")
	}

	return(config)
}

# Remove rows from "data" that are not from one of the 
# designated GWAS.
filter_by_gwas = function(data, gwas, keep=TRUE)
{
	new_data = data[keep == (data$gwas_file %in% gwas),]
	return(new_data)
}

# Remove rows from "data" that are not from one of the 
# designated eQTL tissues.
filter_by_eqtl = function(data, eqtl, keep=TRUE)
{
	new_data = data[keep == (data$eqtl_file %in% eqtl),]
	return(new_data)
}

# create a function that tests GWAS pval / eQTL trait rows for validity
pval_passing = function(x, threshold_set)
{
	pvalue = as.numeric(x[1])
	trait = unlist(x[2])
	if (pvalue < threshold_set$standard)
	{
		return(TRUE)
	}
	else if (("exceptions" %in% names(threshold_set)) && (trait %in% names(threshold_set$exceptions)))
	{
		if (pvalue < threshold_set$exceptions[[trait]])
		{
			return(TRUE)
		}
	}
	return(FALSE)
}

load_results_file = function(config)
{
	if ((("post-hoc-filter") %in% names(config)) && ("full_results_file" %in% names(config$post-hoc-filter)))
	{
		return(read.table(file=config$post-hoc-filter$full_results_file, header=TRUE))
	}
	else
	{
		return(read.table(file=paste(config$out_dir, "raw_coloc_table.txt", sep="/")),header=TRUE)
	}
}

# Function that tests GWAS pval / eQTL trait rows for validity
pval_passing = function(x, threshold_set)
{
	pvalue = as.numeric(x[1])
	trait = unlist(x[2])
	if (pvalue < threshold_set$standard)
	{
		return(TRUE)
	}
	else if (("exceptions" %in% names(threshold_set)) && (trait %in% names(threshold_set$exceptions)))
	{
		if (pvalue < threshold_set$exceptions[[trait]])
		{
			return(TRUE)
		}
	}
	return(FALSE)
}


apply_pval_filter = function(results, config)
{
	filtered_results = results
	if ("gwas_pval_threshold" in config$post-hoc-filter)
	{
		filtered_results = filtered_results[apply(filtered_results[c("pvalue", "trait")], 1, FUN=pval_passing, threshold = config$post-hoc-filter$gwas_pval_threshold),] 
	}
	if ("eqtl_pval_threshold" in config$post-hoc-filter)
	{
		filtered_results = filtered_results[apply(filtered_results[c("pvalue", "trait")], 1, FUN=pval_passing, threshold = config$post-hoc-filter$eqtl_pval_threshold),] 
	}

	return(filtered_results)
}

main()
